File: 백준/Gold/4803. 트리/트리.cc
#include <iostream>

using namespace std;

const int MX = 502;

int n, m;
int a, b;
int t = 0;

int par[MX];
int sz[MX];

int find(int x){
    if(par[x] == x) return x;
    else return par[x] = find(par[x]);
}

void union_(int a, int b){
    a = find(a);
    b = find(b);
    
    if(a==b) return;
    
    sz[b] += sz[a];
    par[a] = b;
}


void solution(){
    for(int i = 1; i <= n; i++) {
        par[i] = i;
        sz[i] = 1;
    }
    
    for(int i = 0; i < m; i++){
        cin >> a >> b;
        a = find(a);
        b = find(b);
        if(a == b) {
          sz[a] = -1e5;
          sz[b] = -1e5;
        }
        union_(a,b);
    }
    
    int cnt = 0;
    
    for(int i = 1; i <= n; i++){
        if(sz[find(i)] <= 0) continue;
        
        sz[find(i)] = 0;
        cnt++;
    }
    
    if(cnt == 0) cout << "No trees.";
    else if(cnt == 1) cout << "There is one tree.";
    else cout << "A forest of " << cnt << " trees.";
    cout << '\n';
}

int main()
{
  
    while(true){
        t++;
        cin >> n >> m;
 
        if(n == 0 && m == 0) break;
        cout << "Case " << t << ": ";
        solution();
    };
    
    return 0;
}

File: 백준/Gold/4803. 트리/트리.java
import java.io.*;
import java.util.*;

public class Main {
	static final int MX = 500;
	static int par[] = new int[MX + 1];

	static void setup() {
		for (int i = 0; i < MX + 1; i++) {
			par[i] = i;
		}
	}

	static int find(int a) {
		if (a == par[a])
			return a;

		return par[a] = find(par[a]);
	}

	static void union(int a, int b) {
		
		a = find(a);
		b = find(b);
		if(a == 0 || b == 0) {
		    par[a] = par[b] = 0;
			return;
		}
		if (a == b) {
			par[b] = 0;
			return;
		}

		par[b] = a;
	}

	public static void main(String[] args) throws Exception {
		// System.setIn(new FileInputStream("src/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		StringBuilder sb = new StringBuilder();
		int c = 1;
		while (true) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			int N = Integer.parseInt(st.nextToken());
			int M = Integer.parseInt(st.nextToken());
			if (N == 0 && M == 0)
				break;

			setup();

			for (int i = 0; i < M; i++) {
				st = new StringTokenizer(br.readLine());
				int a = Integer.parseInt(st.nextToken());
				int b = Integer.parseInt(st.nextToken());

				union(a, b);
			}

			for (int i = 1; i <= N; i++) {
				find(i);
			}

			Set<Integer> set = new HashSet<>();
			for (int i = 1; i <= N; i++) {
				if (par[i] == 0)
					continue;
				set.add(par[i]);
			}

			sb.append("Case ").append(c++).append(": ");
			int cnt = set.size();
			if (cnt == 1) {
				sb.append("There is one tree.\n");
			} else if (cnt == 0) {
				sb.append("No trees.\n");
			} else {
				sb.append("A forest of ").append(cnt).append(" trees.\n");
			}

		}
		System.out.println(sb);
	}

}